python/ray/tests/test_traceback.py

```diff
@@ -0,0 +1,271 @@
+import re
+import sys
+
+import pytest
+import ray
+
+from ray.exceptions import RayTaskError, RayActorError
+"""This module tests stacktrace of Ray.
+
+There are total 3 different stacktrace types in Ray.
+
+1. Not nested task (including actor creation) or actor task failure.
+2. Chained task + actor task failure.
+3. Dependency failure (upstreamed dependency raises an exception).
+
+There are important factors.
+- The root cause of the failure should be printed at the bottom.
+- Ray-related code shouldn't be printed at all to the user-level stacktrace.
+- It should be easy to follow stacktrace.
+
+Each of test verifies that there's no regression by comparing the line number.
+If we include unnecessary stacktrace (e.g., logs from internal files),
+these tests will fail.
+"""
+
+
+def scrub_traceback(ex):
+    assert isinstance(ex, str)
+    ex = ex.strip("\n")
+    ex = re.sub("pid=.*,", "pid=XXX,", ex)
+    ex = re.sub("ip=.*\)", "ip=YYY)", ex)
+    ex = re.sub("line .*,", "line ZZ,", ex)
+    ex = re.sub('".*"', '"FILE"', ex)
+    # These are used to coloring the string.
+    ex = re.sub("\\x1b\[36m", "", ex)
+    ex = re.sub("\\x1b\[39m", "", ex)
+    # noqa is required to ignore lint, so we just remove it.
+    ex = re.sub(" # noqa", "", ex)
+    # Clean up object address.
+    ex = re.sub("object at .*>", "object at ADDRESS>", ex)
+    return ex
+
+
+@pytest.mark.skipif(
+    sys.platform == "win32",
+    reason="Clean stacktrace not supported on Windows")
+def test_actor_creation_stacktrace(ray_start_regular):
+    """Test the actor creation task stacktrace."""
+    expected_output = """The actor died because of an error raised in its creation task, ray::A.__init__() (pid=55184, ip=192.168.1.5) # noqa
+  File "/Users/sangbincho/work/ray/python/ray/tests/test_traceback.py", line 50, in __init__
+    g(3)
+  File "/Users/sangbincho/work/ray/python/ray/tests/test_traceback.py", line 45, in g
+    raise ValueError(a)
+ValueError: 3"""
+
+    def g(a):
+        raise ValueError(a)
+
+    @ray.remote
+    class A:
+        def __init__(self):
+            g(3)
+
+        def ping(self):
+            pass
+
+    try:
+        a = A.remote()
+        ray.get(a.ping.remote())
+    except RayActorError as ex:
+        print(ex)
+        assert scrub_traceback(expected_output) == scrub_traceback(str(ex))
+
+
+@pytest.mark.skipif(
+    sys.platform == "win32",
+    reason="Clean stacktrace not supported on Windows")
+def test_task_stacktrace(ray_start_regular):
+    """Test the normal task stacktrace."""
+    expected_output = """ray::f() (pid=57290, ip=192.168.1.5) # noqa
+  File "/Users/sangbincho/work/ray/python/ray/tests/test_traceback.py", line 96, in f
+    return g(c)
+  File "/Users/sangbincho/work/ray/python/ray/tests/test_traceback.py", line 88, in g
+    raise ValueError(a)
+ValueError: 7"""
```

Comment:
Should probably commit post scrubbed output here, to avoid regex accidentally replacing everything bugs.

Reply:
Oh yeah good point. I will modify that